### ISSUE TITLE
Rule proposal: `prefer-instanceof`

### ISSUE THREAD

**Issue by kidonng** [sindresorhus/eslint-plugin-unicorn#1574]
Checking the type of an element via `element.tagName === 'name'` doesn't give you type guard and you have to use optional chain (or `!` a.k.a. Typescript's non-null assertion operator) if the element can be `undefined` or `null`.

An real life example is: [`574d442` (#5036)](https://github.com/refined-github/refined-github/pull/5036/commits/574d442d6ac794621624bb59c94581e56da392f3)

## Fail

```ts
if (event.target.parentElement?.tagName === 'img') {
	return (target as HTMLImageElement).src
}
```

## Pass

```js
if (event.target.parentElement instanceof HTMLImageElement) {
	return target.src
}
```

**Comment by sindresorhus** [comment]
I think the rule scope and name should be more narrow. `instanceof` is generally not the best solution for most type-checks.

This will also most likely get solved in TypeScript at some point: https://github.com/microsoft/TypeScript/pull/38839

**Comment by kidonng** [comment]
> I think the rule scope and name should be more narrow. `instanceof` is generally not the best solution for most type-checks.

How about `prefer-instanceof-element`? 

> This will also most likely get solved in TypeScript at some point: [microsoft/TypeScript#38839](https://github.com/microsoft/TypeScript/pull/38839)

I'm aware of that but it will probably take forever. And there are benefits outside TS too.

**Comment by sindresorhus** [comment]
I would name it `prefer-instanceof-dom-element` since `element` is a general term, for example, an element in a collection.

**Comment by sindresorhus** [comment]
> I'm aware of that but it will probably take forever. And there are benefits outside TS too.

What are the non-TS benefits?

**Comment by sindresorhus** [comment]
I believe this was fixed in TS, so closing.
